typing/printtyp.ml

```diff
@@ -1527,49 +1527,35 @@ let extension_constructor_args_and_ret_type_subtree ext_args ext_ret_type =
   let args = tree_of_constructor_arguments ext_args in
   (args, ret)
 
-let isolated_extension_constructor_args_and_ret_type_subtree
-    ext_args ext_ret_type =
-  match ext_ret_type with
-  | None ->
-      extension_constructor_args_and_ret_type_subtree ext_args ext_ret_type
-  | Some _ ->
-      Names.with_local_names
-        (fun () ->
-           extension_constructor_args_and_ret_type_subtree
-             ext_args ext_ret_type
-        )
-
-(* When printing extension constructor, we need to forget the naming decisions
-   of different parts of the extension constructor depending on the context.
-   For example, in
-   {[
-     type 'b t += A: 'b -> 'b any t
-   ]}
-   We can choose to either use local names for the right hand GADT constructor
-   (Isolated_gadt), or the left hand type parameters (Isolated_param).
-   Isolating the gadt constructor is the same behaviour as
-   [isolated_tree_of_constructor].
-   We would choose to isolate the type parameters if we want to refer to the
-   GADT variables from another part of the printing context.
-   *)
-type extension_constructor_scoping =
-  | Isolated_gadt
-  | Isolated_param
-
-let add_extension_constructor_to_preparation scoping ext =
-  begin match scoping with
-  | Isolated_gadt ->
-    let ty_params = filter_params ext.ext_type_params in
-    List.iter add_alias ty_params;
-    List.iter prepare_type ty_params;
-    List.iter add_printed_alias ty_params;
-  | Isolated_param -> ()
-  end;
+(* When printing extension constructor, it is important to ensure that
+after printing the constructor, we are still in the scope of the constructor.
+For GADT constructor, this can be done by printing the type parameters inside
+their own isolated scope. This ensures that in
+{[
+   type 'b t += A: 'b -> 'b any t
+]}
+the type parameter `'b` is not bound when printing the type variable `'b` from
+the constructor definition from the type parameter.
+
+Contrarily, for non-gadt constructor, we must keep the same scope for
+the type parameters and the constructor because a type constraint may
+have changed the name of the type parameter:
+{[
+type -'a t = .. constraint <x:'a. 'a t -> 'a> = 'a
+(* the universal 'a is here to steal the name 'a from the type parameter *)
+type 'a t = X of 'a
+]} *)
```

Comment:
(This is not a change request, just a remark I had while reading the comment.)

I wondered about what happens for types that contain both constraints and GADTs, but apparently currently constraints are ignored for GADT constructors:

```ocaml
type 'a t = 
  | Any : 'a -> int t
  | Int of int
  constraint 'a = int
(*
Warning 62 [constraint-on-gadt]: Type constraints do not apply to GADT cases of variant types.
*)
```

(I'm not sure what "do not apply" means here as the constraints are still checked, for example making Any return a `unit t` fails with an error. Oh well.)